extras/sht3x/sht3x.c

```diff
@@ -0,0 +1,419 @@
+
+/** Forward declaration of function for internal use */
+
+static bool sht3x_send_command  (sht3x_sensor_t*, uint16_t);
+static bool sht3x_read_data     (sht3x_sensor_t*, uint8_t*,  uint32_t);
+static bool sht3x_get_status    (sht3x_sensor_t*, uint16_t*);
+// static bool sht3x_reset         (sht3x_sensor_t*);
+
+static uint8_t crc8 (uint8_t data[], int len);
+
+/** ------------------------------------------------ */
+
+bool sht3x_init_driver()
+{
+    return true;
+}
+
+
+sht3x_sensor_t* sht3x_init_sensor(uint8_t bus, uint8_t addr)
+{
+    sht3x_sensor_t* dev;
+    
+    if ((dev = malloc (sizeof(sht3x_sensor_t))) == NULL)
+        return NULL;
+    
+    // init sensor data structure
+    dev->bus  = bus;
+    dev->addr = addr;
+    dev->mode = single_shot;
+    dev->repeatability = high;
+    dev->meas_started = false;
+    dev->meas_start_time = 0;
+    
+    dev->active = true;
+
+    uint16_t status;
+
+    // check the again the status after clear status command
+    if (!sht3x_get_status(dev, &status))
+    {
+        error_dev ("could not get sensor status", __FUNCTION__, dev);
+        free(dev);
+        return NULL;       
+    }
+    
+    debug_dev ("sensor initialized", __FUNCTION__, dev);
+    return dev;
+}
+
+
+int32_t sht3x_start_measurement (sht3x_sensor_t* dev, sht3x_mode_t mode)
+{
+    if (!dev || !dev->active) return -1;
+    
+    dev->error_code = SHT3x_OK;
+    
+    // return remaining time when measurement is already running
+    if (dev->meas_started)
+    {
+        debug_dev ("measurement already started", __FUNCTION__, dev);
+        dev->error_code = SHT3x_MEAS_ALREADY_RUNNING;
+        return sht3x_is_measuring (dev);
+    }
+    
+    dev->mode = mode;
+    
+    // start measurement according to selected mode and return an duration estimate
+    if (sht3x_send_command(dev, SHT3x_MEASURE_CMD[dev->mode][dev->repeatability]))
+    {
+        dev->meas_start_time = sdk_system_get_time ();
+        dev->meas_started = true;
+        dev->meas_first = true;
+        return SHT3x_MEASURE_DURATION[dev->repeatability];   // in ms
+    }
+
+    dev->error_code |= SHT3x_SEND_MEAS_CMD_FAILED;
+    return -1;  // on error
+}
+
+
+// returns 0 when finished, -1 on error, remaining measurement time otherwise
+int32_t sht3x_is_measuring (sht3x_sensor_t* dev)
+{
+    if (!dev || !dev->active) return -1;
+
+    dev->error_code = SHT3x_OK;
+
+    if (!dev->meas_started)
+    {
+        error_dev ("measurement not started", __FUNCTION__, dev);
+        dev->error_code = SHT3x_MEAS_NOT_STARTED;
+        return -1;
+    }
+    
+    // computation is necessary in periodic mode only for first measurement
+    // in single shot mode every measurment is the first one
+    if (!dev->meas_first)
+        return 0;
+        
+    uint32_t elapsed_time;
+
+    elapsed_time = (sdk_system_get_time() - dev->meas_start_time) / 1000;  // in ms
```

Comment:
Just a thought, but might `xTaskGetTickCount()` be better used for timing here. The measurement durations might then be in ticks too, and then the integer division could be avoided. Also the number of ticks to wait would be returned and could be used by the caller to pass directly to `vTaskDelay()` without another division.

Reply:
Good idea, makes really sense.